### DISCUSSION TITLE
how do i change the naming scheme when creating a new file.

### DISCUSSION THREAD

**Discussion by shaw4m4n** [epwalsh/obsidian.nvim#295] [Q&A] [answered]
when you create a file it uses random name. you have to use .md with the name to use heading as file name. how can I create a file with heading name as file name without writing .md in links?

**Comment by epwalsh** [comment, marked as answer]
Hey @shawaman, in your configuration you can specify a `note_id_func`. This is just a Lua function that's responsible for determining the ID/filename of new notes based on the title. The function I use personally looks like this:

```lua
  note_id_func = function(title)
    -- Create note IDs in a Zettelkasten format with a timestamp and a suffix.
    -- In this case a note with the title 'My new note' will be given an ID that looks
    -- like '1657296016-my-new-note', and therefore the file name '1657296016-my-new-note.md'
    local suffix = ""
    if title ~= nil then
      -- If title is given, transform it into valid file name.
      suffix = title:gsub(" ", "-"):gsub("[^A-Za-z0-9-]", ""):lower()
    else
      -- If title is nil, just add 4 random uppercase letters to the suffix.
      for _ = 1, 4 do
        suffix = suffix .. string.char(math.random(65, 90))
      end
    end
    return tostring(os.time()) .. "-" .. suffix
  end,
```

If you literally want the title/heading as the note name you could simply return `title`.

**Reply by shaw4m4n** [reply]
I tried this before. which changed the behavior for :ObsidianNew command as intended. but creating filename by editing in files and linking stopped working. Completions and other shortcuts to enter file like gf and CR also stooped working 

` note_id_func = function(title)
                return tostring(title)
            end,
`

**Reply by shaw4m4n** [reply]
how can i modify the behaviour of cmp so it can work with simple file names?

**Reply by epwalsh** [reply]
It seems to work fine on my end, so it sounds like you're running into a bug. Can you open an issue with more details on how to reproduce?

**Reply by rodhash** [reply]
Apparently this one still adds a random number into the filename .. it also adds the title as a suffix but still adds a random number

The below solution from @shawaman seems to work fine.

**Comment by shaw4m4n** [comment]
It's working fine now with the following.
```lua
note_id_func = function(title)
                local suffix = ""
                if title ~= nil then
                    suffix = title:gsub(" ", "-"):gsub("[^A-Za-z0-9-]", ""):lower()
                else
                    for _ = 1, 4 do
                        suffix = suffix .. string.char(math.random(65, 90))
                        suffix = tostring(os.time()) .. "-" .. suffix
                    end
                end
                return suffix
            end,
```

**Reply by rodhash** [reply]
Yep this one worked fine
